lib/Jsrt/JsrtDebuggerObject.cpp

```diff
@@ -808,43 +822,51 @@ Js::JavascriptArray * JsrtDebugStackFrames::StackFrames(Js::ScriptContext * scri
+
+        // Walk all the script contexts and collect the frames which are below the address when break was reported.
+        for (Js::ScriptContext *tempScriptContext = threadContext->GetScriptContextList();
+        tempScriptContext != nullptr && tempScriptContext->IsScriptContextInDebugMode();
+            tempScriptContext = tempScriptContext->next)
         {
-            Js::DiagStack* stackFrames = framePointers->GetStrongReference();
-            if (stackFrames != nullptr)
+            Js::WeakDiagStack * framePointers = tempScriptContext->GetDebugContext()->GetProbeContainer()->GetFramePointers(dispatchHaltFrameAddress);
+            if (framePointers != nullptr)
             {
-                int count = stackFrames->Count();
-                for (int frameIndex = 0; frameIndex < count; ++frameIndex)
+                Js::DiagStack* stackFrames = framePointers->GetStrongReference();
+                if (stackFrames != nullptr)
                 {
-                    Js::DiagStackFrame* stackFrame = stackFrames->Peek(frameIndex);
-
-                    DWORD_PTR stackAddress = stackFrame->GetStackAddress();
-                    // Only give frames which were preset when break was triggered as we might have other user (non-library) scripts executed after break
-                    if (stackAddress >= dispatchHaltFrameAddress)
+                    int count = stackFrames->Count();
+                    for (int frameIndex = 0; frameIndex < count; ++frameIndex)
                     {
-                        Js::DynamicObject* stackTraceObject = this->GetStackFrame(stackFrame, frameCount);
-
-                        Js::Var marshaledObj = Js::CrossSite::MarshalVar(scriptContext, stackTraceObject);
-                        Js::JavascriptOperators::OP_SetElementI((Js::Var)stackTraceArray, Js::JavascriptNumber::ToVar(frameCount++, scriptContext), marshaledObj, scriptContext);
+                        Js::DiagStackFrame* stackFrame = stackFrames->Peek(frameIndex);
+                        stackList->Add(stackFrame);
                     }
                 }
-            }
 
-            framePointers->ReleaseStrongReference();
-            HeapDelete(framePointers);
+                framePointers->ReleaseStrongReference();
+                HeapDelete(framePointers);
+            }
         }
+
+        // Frames can be from multiple contexts, sort them based on stack address
+        stackList->Sort(DiagStackFrameSorter, (void*)dispatchHaltFrameAddress);
+
+        stackList->Map([&](int index, Js::DiagStackFrame* stackFrame)
+        {
+            Js::DynamicObject* stackTraceObject = this->GetStackFrame(stackFrame, index);
+            Js::Var marshaledObj = Js::CrossSite::MarshalVar(scriptContext, stackTraceObject);
+            Js::JavascriptOperators::OP_SetElementI((Js::Var)stackTraceArray, Js::JavascriptNumber::ToVar(index, scriptContext), marshaledObj, scriptContext);
```

Comment:
nit: we could create the stackTraceArray just before this Map function? and also CreateArray takes the length as well, as we know by this time the actual length we could create the array the actual length.